File: 42.trapping_rain_water.go
package main

func trap(height []int) int {
	count := 0

	for i := 0; i < len(height); i++ {
		var left int
		var right int
		cur := height[i]

		for m := i - 1; m > -1; m-- {
			if height[m] > cur && height[m] > left {
				left = height[m]
			}
		}

		for n := i + 1; n < len(height); n++ {
			if height[n] > cur && height[n] > right {
				right = height[n]
			}
		}

		if left > 0 && right > 0 {
			min := 0
			if left > right {
				min = right
			} else {
				min = left
			}
			count = count + min - cur
		}
	}

	return count
}

File: 42.trapping_rain_water.ts
function trap(height: number[]): number {
  let count: number = 0;
  const leftMax: number[] = [];
  const rightMax: number[] = [];

  for (let i = 1; i < height.length; i++) {
    if (i === 1) {
      const [firstHeight] = height;
      leftMax[i] = firstHeight;
    } else {
      leftMax[i] = Math.max(leftMax[i - 1], height[i - 1]);
    }
  }

  for (let i = height.length - 2; i >= 0; i--) {
    if (i === height.length - 2) {
      rightMax[i] = height[height.length - 1];
    } else {
      rightMax[i] = Math.max(rightMax[i + 1], height[i + 1]);
    }
  }

  for (let i = 1; i < height.length - 1; i++) {
    if (leftMax[i] > height[i] && rightMax[i] > height[i]) {
      count = count + Math.min(leftMax[i], rightMax[i]) - height[i];
    }
  }

  return count;
}
